Scripts/Duel: implement duel reset of health and mana and improve duel reset

FrancescoBorzi opened the pull request:
This PR is about the Duel Reset script and it is composed by several parts:
## Implementation of Health & Mana reset

Originally written by @GigaDev90 (https://github.com/TrinityCore/TrinityCore/pull/15728), rewritten using scripts.

This PR introduces a configurable feature to reset health and mana before and after duel.
- Before duel it saves health and mana(if any), and set them to the max value
- After duel it restore the old health and mana(if any)
## Improvement of the whole Duel Reset system

This PR also introduces a global improvement for the Duel Reset system: the reset should not be performed if duel isn't started (DUEL_INTERRUPTED) or fled (DUEL_FLED).
## Fix restoration of the Druid mana

The initial part didn't handle the restoration of the Druid mana, with a commit of Giga we handle that too. (by @GigaDev90)
## Fix onHold spells

Fixed bug when a player accept duel with a spell on onHold true (like sthealt of dudu when is active). Now in any condition of onHold never happens visual bug or wrong store of cooldown. (by @GigaDev90)

FrancescoBorzi committed:
Scripts/DuelReset:
- impemented health and mana reset
- do not reset anything when duel is interrupted

joschiwald commented on src/server/game/Entities/Player/Player.h:
```diff
@@ -1946,6 +1946,12 @@ class Player : public Unit, public GridObject<Player>
+        void SaveHealthBeforeDuel() { healthBeforeDuel = GetHealth(); }
+        void SaveManaBeforeDuel() { manaBeforeDuel = GetPower(POWER_MANA); }
+        void RestoreHealthAfterDuel() { SetHealth(healthBeforeDuel); }
+        void RestoreManaAfterDuel() { SetPower(POWER_MANA, manaBeforeDuel); }
```
you could save this stuff in script too

Shauren replied:
no, you cant. PlayerScript is not the same as an AI script - there is only one global instance of it

joschiwald replied:
... by using a std::unordered_map with guid as key

ShinDarth replied:
I think the current one is the most elegant way....

Shauren replied:
@joschiwald which would be again global - and not threadsafe

Naios replied:
Maybe we could add a "custom" store in player which is adaptable but not global.
We could use it to replace several other less used optional core or custom stores,
which would also increase the compression factor for unused entries:

```c++
enum class PlayerExtraUINT32Store
{
    STORE_PLAYER_HEALTH_BEFORE_DUEL,
    STORE_PLAYER_MANA_BEFORE_DUEL
};

Optional<uint32> Player::GetCustomStorageEntry(PlayerExtraUINT32Store::STORE_PLAYER_MANA_BEFORE_DUEL)
{
    ...
}

bool Player::ClearCustomStorageEntry(PlayerExtraUINT32Store::STORE_PLAYER_MANA_BEFORE_DUEL)
{
    ...
}
```

Shauren replied:
Optional still takes memory, even if not initialized

Naios replied:
No... the value would be stored in a hashmap... The optional is only the return type

Shauren replied:
Then again, it doesnt have to be an uint32 (hence boost::any suggestion)

Naios replied:
Right

GigaDev90 committed:
Scripts/DuelReset:
- fixed druid mana restoration
- fixed bug when a player accepts duel with a spel on onHold true (like when stealth of rogue/druid is active)

ShinDarth commented:
I would like to avoid commit squashing, since there are commits doing different things and/or with different authors

FrancescoBorzi committed:
Scripts/DuelReset:
- do not reset when duel is fled
- minor improvement to SpellHistory::RestoreCooldonStateAfter duel method

ShinDarth commented:
This PR fixed all bugs of the duel reset system and added new features to it.

I've been testing it for a while, and it seems to work fine. There is no reason for this PR to be still open, please merge it

tkrokli commented:
Nice job, ShinDarth! :+1:
